Add support for GIMP and JASC palettes (with per-color alpha)

reaperrr opened the pull request:
Supersedes #13913.

Compared to #13913, following changes and additions were made:
- merged GIMP/JASC support into a single trait
- replaced the limited-use `Tileset` with `Tilesets` + `ExcludeTilesets`
- Added support for per-color alpha
- updated testcase commits to account for per-color alpha (RA effects, TS units) and make testing quicker

Mailaender committed:
Add support for JASC and GIMP color palettes

Mailaender committed:
JASC TESTCASE

Mailaender committed:
GIMP TESTCASE

Phrohdoh commented on mods/cnc/rules/palettes.yaml:
```diff
@@ -23,6 +23,10 @@
 		Filename: temperat.pal
 		ShadowIndex: 4
 		AllowModifiers: false
+	PaletteFromGimpFile:
```
This needs to be `PaletteFromGimpOrJascFile`, right?

reaperrr replied:
Stupid me, of course.

Phrohdoh commented:
There are quite a few `PaletteFromGimpFile`s in yaml that need to be `PaletteFromGimpOrJascFile`.

reaperrr commented:
Updated.

GraionDilach reviewed (approved):
Cooooool. :+1:

pchote commented:
Do you want to merge the example commit, or is that just a testcase? (IMO they belong in https://github.com/OpenRA/ArtSrc, not here)

pchote commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs:
```diff
@@ -0,0 +1,131 @@
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
```
Modders can define shadows (including alpha) directly in the palette, so this is redundant.

pchote commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs:
```diff
@@ -0,0 +1,131 @@
+
+						if (r == 0 && g == 0 && b == 0)
+							colors[i] = 0;
+						else if (a == 0)
```
Doesn't this prevent fully transparent colours?

pchote replied:
The issues here could be solved by treating `r == 0 && g == 0 && b == 0 && a == 0` (transparent) as distinct from `r == 0 && g == 0 && b == 0` (black)

reaperrr replied:
But alpha of 0 means zero transparency...?

pchote replied:
Sorry yes, I meant non-transparent black on the case above.

reaperrr commented:
> Do you want to merge the example commit?

No, I'll rename the commit titles to make that more clear.

reaperrr commented:
Updated.

Palette index 0 is now always converted to fully transparent regardless of original color, in favor of otherwise keeping the code as simple and index-hack-free as possible.
Removed ShadowIndex in favor of using a black or dark-grey color + alpha.

Also updated testcases:
- GIMP testcase is RA, with explosion effects showcasing per-color alpha (transparency increases the darker the explosion color gets).
- JASC testcase is TS, showcasing non-back index 0 being converted to transparent, black + 128 alpha serving as shadow (color index 1).

reaperrr commented:
Rebased to fix the AppVeyor issue.

reaperrr commented:
@GraionDilach is your +1 still valid after the updates?

GraionDilach commented:
I'm not happy with the slot 0 aspect, Photoshop throws the unused color slots to the end of it's palette when saving an indexed image, which might throw people off.

GraionDilach commented:
On a second thought, I'm fine with an explicit mention of the slot 0 hardcode in the trait description.

reaperrr commented:
OK. We can always update the trait if it turns out to be an issue.

GraionDilach reviewed (approved):
:+1:

Mailaender commented:
Cherry-picked to bleed without the test cases https://github.com/OpenRA/OpenRA/commit/12054506e1399bccadd94727299266fc3133821f.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/29d4d2961f33be061eafde10e518ad4f57789a80...bcf8c022dacc0208dad4bd02db211f9bbbd42da6)
